spark3/src/main/java/org/apache/iceberg/actions/Spark3MigrateAction.java

```diff
@@ -0,0 +1,144 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Some;
+import scala.collection.JavaConverters;
+
+/**
+ * Takes a Spark table in the sourceCatalog and attempts to transform it into an Iceberg
+ * Table in the same location with the same identifier. Once complete the identifier which
+ * previously referred to a non-iceberg table will refer to the newly migrated iceberg
+ * table.
+ */
+class Spark3MigrateAction extends Spark3CreateAction {
+  private static final Logger LOG = LoggerFactory.getLogger(Spark3MigrateAction.class);
+  private static final String BACKUP_SUFFIX = "_BACKUP_";
+
+  Spark3MigrateAction(SparkSession spark, CatalogPlugin sourceCatalog, Identifier sourceTableName) {
+    super(spark, sourceCatalog, sourceTableName, sourceCatalog, sourceTableName);
+  }
+
+  @Override
+  public Long execute() {
+    Table icebergTable;
+
+    // Move source table to a new name, halting all modifications and allowing us to stage
+    // the creation of a new Iceberg table in its place
+    String backupName = sourceTableName().name() + BACKUP_SUFFIX;
+    Identifier backupIdentifier = Identifier.of(sourceTableName().namespace(), backupName);
+    try {
+      destCatalog().renameTable(sourceTableName(), backupIdentifier);
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException("Cannot find table to migrate", e);
+    } catch (TableAlreadyExistsException e) {
+      throw new IllegalArgumentException("Cannot rename migration source to backup name", e);
+    }
+
+    StagedTable stagedTable = null;
+    try {
+      Map<String, String> newTableProperties = new ImmutableMap.Builder<String, String>()
+          .put(TableCatalog.PROP_PROVIDER, "iceberg")
+          .putAll(JavaConverters.mapAsJavaMapConverter(v1SourceTable().properties()).asJava())
+          .putAll(tableLocationProperties(sourceTableLocation()))
+          .putAll(additionalProperties())
+          .build();
+
+      stagedTable = destCatalog().stageCreate(Identifier.of(
+          destTableName().namespace(),
+          destTableName().name()),
+          v1SourceTable().schema(),
+          sourcePartitionSpec(),
+          newTableProperties);
+
+      icebergTable = ((SparkTable) stagedTable).table();
+
+      if (!icebergTable.properties().containsKey(TableProperties.DEFAULT_NAME_MAPPING)) {
+        applyDefaultTableNameMapping(icebergTable);
+      }
+
+      String stagingLocation = icebergTable.location() + "/" + ICEBERG_METADATA_FOLDER;
+
+      LOG.info("Beginning migration of {} using metadata location {}", sourceTableName(), stagingLocation);
+
+      Some<String> backupNamespace = Some.apply(backupIdentifier.namespace()[0]);
+      TableIdentifier v1BackupIdentifier = new TableIdentifier(backupIdentifier.name(), backupNamespace);
+      SparkTableUtil.importSparkTable(spark(), v1BackupIdentifier, icebergTable, stagingLocation);
+
+      stagedTable.commitStagedChanges();
+    } catch (Exception e) {
```

Comment:
I think it is usually better to do work like this in a `finally` block using a `threw` boolean:

```java
boolean threw = true;
try {
  // something that might fail
  threw = false;
} finally {
  if (threw) {
    // clean up
  }
}
```

That has the advantage that whatever happened in the block, you get the correct exception rather than needing to throw a generic `RuntimeException` which would make it difficult to use this action in a higher-level application. Also, there are technically throwables that are not `Exception`.

Reply:
Sounds good, i'll change it around